File: Platforms/Win32/Reign.Video.D3D9/Shaders/VertexShader.cs
using System;
using Reign.Core;
using Reign_Video_D3D9_Component;

namespace Reign.Video.D3D9
{
	public class VertexShader : ShaderModel
	{
		#region Properties
		private VertexShaderCom vertexShaderCom;
		#endregion

		#region Constructors
		public VertexShader(DisposableI parent, string code, ShaderVersions shaderVersion)
		: base(parent, code, shaderVersion, ShaderTypes.VS)
		{
			try
			{
				var video = parent.FindParentOrSelfWithException<Video>();

				vertexShaderCom = new VertexShaderCom();
				var error = vertexShaderCom.Init(video.com, com);

				switch (error)
				{
					case VertexShaderErrors.VertexShader: Debug.ThrowError("VertexShader", "Failed to create vertex shader"); break;
				}
			}
			catch (Exception e)
			{
				Dispose();
				throw e;
			}
		}

		public override void Dispose()
		{
			disposeChilderen();
			if (vertexShaderCom != null)
			{
				vertexShaderCom.Dispose();
				vertexShaderCom = null;
			}
			base.Dispose();
		}
		#endregion

		#region Methods
		public void Apply()
		{
			vertexShaderCom.Apply();
		}
		#endregion
	}
}

File: Platforms/Win32/Reign_Video_D3D9_Component/VertexShader.cpp
#include "pch.h"
#include "VertexShader.h"

namespace Reign_Video_D3D9_Component
{
	#pragma region Constructors
	VertexShaderErrors VertexShaderCom::Init(VideoCom^ video, ShaderModelCom^ shaderModel)
	{
		null();
		this->video = video;

		IDirect3DVertexShader9* shaderTEMP = 0;
		// TODO: load data file file
		/*if (FAILED(video->device->CreateVertexShader((DWORD*)shaderModel->code->GetBufferPointer(), &shaderTEMP)))
		{
			return VertexShaderErrors::VertexShader;
		}*/
		this->shader = shaderTEMP;

		return VertexShaderErrors::None;
	}

	VertexShaderCom::~VertexShaderCom()
	{
		if (shader) shader->Release();
		null();
	}

	void VertexShaderCom::null()
	{
		shader = 0;
	}
	#pragma endregion

	#pragma region Methods
	void VertexShaderCom::Apply()
	{
		video->device->SetVertexShader(shader);
	}
	#pragma endregion
}